Allow SQLT options to be passed to unique constraints

Altreus opened the pull request:
Recreated #87 because I based it off master instead of blead.

This PR does work against the current version of SQLT, but that version of SQLT doesn't actually do anything with the information.

There is a PR on SQLT that will actually cause this to have some effect.

dbsrgits/sql-translator#75

I will try to write some tests for this, but I wanted to get the PR made so I can invite discussion in the meantime.

ribasushi commented:
Greetings @Altreus 

Sigh... this is an unholy shitpile (the API before your changes, not your fault). Currently we have:

`add_unique_constraint`
`add_unique_constraints`
`name_unique_constraint`
`unique_constraints`
`unique_constraint_names`
`unique_constraint_columns`

Adding to this `unique_constraints_extra` to this clusterfuck is not wise to say the least. Instead please add a `->unique_constraints_info` method, make it return a hashref (not a flattened list), with names as keys, and values being a subhashref with `columns`, `extra`, etc... fields. Just the way `->columns_info` works.

That'd be the starting point, which invalidates what you currently have, so I can't elaborate further. More once the first mockup is up.

Altreus commented:
I did wonder about creating a single method that returned the whole lot, but I figured that since I'd be adding a new method anyway I should follow suit.

I think you're saying to do what I was originally going to do, so I'll implement that and push the updated branch.

Altreus commented:
Is this what you were thinking of, @ribasushi ?

ribasushi commented on lib/DBIx/Class/ResultSource.pm:
```diff
@@ -702,7 +702,7 @@ sub sequence {
 
-=item Arguments: $name?, \@colnames
+=item Arguments: $name?, \@colnames, \%extra?
 
```
This is not a useful way to extend the signature - too many unknowns. Make it

`$name?, \%uc_info | \@colnames`

With both _names optional_ and _\@colnames_ marked as discouraged in the docs. I.e. people should be doing:

```
add_unique_constraints(
  foo => { ... }
  bar => { ... }
)
```

just like they do with `add_columns` today.

Altreus replied:
This is `add_unique_constraint` rather than `add_unique_constraints`. Shall I also change the plural version to accept hashrefs to follow suit?

This makes more sense, actually, now that I see what you're after.

ribasushi replied:
Along these lines yes. Again - look at how `add_column(s)` does things, model on that.

ribasushi commented on lib/DBIx/Class/ResultSource.pm:
```diff
@@ -724,6 +725,11 @@ This will result in a unique constraint named
 C<table_column1_column2>, where C<table> is replaced with the table
 name.
 
+The "extra" hashref will be passed to
```
It's not an `extra hashref` anymore. User specifies explicit `uc_name => { extra => <what will go to SQLT>, <other options> }`. This way you keep the API extensible in the future, instead of locking things into _all of this will be passed to SQLT and you better hope it interoperates forever, KTHXBAI!_

In fact I would call it `sqlt_extra` (even though there is precedent in https://metacpan.org/pod/DBIx::Class::ResultSource#extra, but that's arguably bad precedent)

ribasushi commented on lib/DBIx/Class/ResultSource.pm:
```diff
@@ -874,7 +889,41 @@ column names as values.
+L</add_unique_constraint> or L</add_unique_constraints>;
+
+C<extra> is the hashref originally passed to
+L</add_unique_constraint>, or an empty hashref if none was provided.
```
Do not do `an empty hashref if none was provided`. Either the user supplied one or they did not, do not add anything on your own.

ribasushi commented on lib/DBIx/Class/ResultSource.pm:
```diff
@@ -874,7 +889,41 @@ column names as values.
+C<extra> is the hashref originally passed to
+L</add_unique_constraint>, or an empty hashref if none was provided.
+
+B<Note> that while similar functions return hash-shaped lists, this
```
`s/hash-shaped/flattened hashes as a list/`, and I also would add `... just like L</columns_info>`

ribasushi commented on lib/SQL/Translator/Parser/DBIx/Class.pm:
```diff
@@ -149,13 +149,16 @@ sub parse {
-            if (!$source->_compare_relationship_keys($unique_constraints{$uniq}, \@primary)) {
+        my $unique_constraints = $source->unique_constraints_info;
+        foreach my $uniq (sort keys %$unique_constraints) {
+            my $cons = $unique_constraints->{$uniq};
```
Don't add extremely short-lived intermediate variables like that, it makes code harder to reason about, as in _why did he save it only to use it a couple times in this function, what am I missing_

Altreus replied:
Surely "because the lines get really long otherwise" is the most common reason?

ribasushi replied:
_Lines get really long_ is exclusively an aesthetic concern. On the other hand _author made the interpreter do extra work... but I can't see why_ is more involved.

Either way - this was a minor nitpick as I was reading. I will fix it when I do the final review/rebase pass.

Ignore this particular comment.

Altreus commented:
@ribasushi I've addressed all your points. Having rewritten `add_unique_constraint` to accept an arrayref or a hashref, I was able to rewrite the docs for `add_unique_constraints` to be along the lines of "It's add_unique_constraint but more than once".

ribasushi commented on lib/DBIx/Class/ResultSource.pm:
```diff
@@ -731,6 +762,9 @@ only columns in the constraint are searched.
 
+Note also that the keys C<table>, C<name>, and C<fields> in
+C<sqlt_extra> will be ignored.
+
```
What does this note actually mean...? Please elaborate

Altreus commented:
Elaborated on ignored keys, as requested

ribasushi commented:
Same here - projecting to look into this in about a week or so. Once again - sorry for the delay. This past month alone has been murder:
$ git diff -w --shortstat 729656c504e..18a2903b824
 389 files changed, 2278 insertions(+), 1138 deletions(-)

Altreus commented:
Just a heads up, I've rebased this because my tests were failing locally for some reason, and blead has moved somewhat.

ribasushi commented:
Nod, very very close now. Sorry for the delays :/

Altreus commented:
No trouble, working from branch locally, hence pesky rebase.

Altreus commented:
Any progress? Sorry to mither but it's holding us back because we're having to install this branch instead of latest

ribasushi commented:
@Altreus terribly sorry you had to sit on this for so long. I just wanted to drop you a line that this is considered a blocker for the current master ( see label ) and won't be dropped on the floor. I will drop you another line once the more immediate maint-issues are taken care of, and dev from within `master` resumes.

Stay tuned, several weeks should be it.
